nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -558,36 +564,46 @@ func (w *deploymentWatcher) handleAllocUpdate(allocs []*structs.AllocListStub) (
 	return res, nil
 }
 
-// shouldRollback returns whether the job should be rolled back to an earlier
-// stable version by examing the allocations in the deployment.
-func (w *deploymentWatcher) shouldRollback() (bool, error) {
+// shouldFail returns whether the job should be failed and whether it should
+// rolled back to an earlier stable version by examing the allocations in the
+// deployment.
+func (w *deploymentWatcher) shouldFail() (fail, rollback bool, err error) {
 	snap, err := w.state.Snapshot()
 	if err != nil {
-		return false, err
+		return false, false, err
 	}
 
 	d, err := snap.DeploymentByID(nil, w.deploymentID)
 	if err != nil {
-		return false, err
+		return false, false, err
 	}
 
+	fail = false
 	for tg, state := range d.TaskGroups {
```

Comment:
Is this map always going to be populated with deployment state for all task groups? i.e can some task groups not be in the map initially and then get added later?

Reply:
No it will only have a mapping for task groups that were being updated at the time of the deployments creation. They are immutable in that sense.